metrique-aggregation/src/histogram.rs

```diff
@@ -0,0 +1,439 @@
+//!
+//!     // Record multiple observations
+//!     metrics.backend_latency.add_value(Duration::from_millis(45));
+//!     metrics.backend_latency.add_value(Duration::from_millis(67));
+//!     metrics.backend_latency.add_value(Duration::from_millis(52));
+//!
+//!     // When metrics drops, emits a single entry with the distribution
+//! }
+//! ```
+//!
+//! # Choosing an aggregation strategy
+//!
+//! By default, histograms use [`ExponentialAggregationStrategy`]. To use a different strategy,
+//! specify it as the second type parameter:
+//!
+//! ```
+//! use metrique_aggregation::histogram::{Histogram, SortAndMerge};
+//! use std::time::Duration;
+//!
+//! let histogram: Histogram<Duration, SortAndMerge> = Histogram::new(SortAndMerge::new());
+//! ```
+//!
+//! ## ExponentialAggregationStrategy (default)
+//!
+//! Uses exponential bucketing with ~6.25% error. This is the best choice for most use cases:
+//!
+//! - Provides consistent relative precision across wide value ranges
+//! - Memory efficient with fixed bucket count (464 buckets)
+//! - Fast recording and draining operations
+//!
+//! Use this when you need good precision across values that span multiple orders of magnitude
+//! (e.g., latencies from microseconds to seconds).
+//!
+//! ## AtomicExponentialAggregationStrategy
+//!
+//! Thread-safe version of exponential bucketing. Use with [`crate::histogram::AtomicHistogram`] when you need
+//! to record values from multiple threads concurrently:
+//!
+//! ```
+//! use metrique_aggregation::histogram::{AtomicHistogram, AtomicExponentialAggregationStrategy};
+//! use std::time::Duration;
+//!
+//! let histogram: AtomicHistogram<Duration, AtomicExponentialAggregationStrategy> =
+//!     AtomicHistogram::new(AtomicExponentialAggregationStrategy::new());
+//! ```
+//!
+//! ## SortAndMerge
+//!
+//! Stores all observations exactly and sorts them on emission:
+//!
+//! - Perfect precision - no bucketing error
+//! - Memory usage grows with observation count
+//! - Slower drain operation due to sorting
+//!
+//! Use this when you need exact values and have a bounded number of observations (typically < 1000).
+
+use metrique_core::CloseValue;
+use metrique_writer::{MetricFlags, MetricValue, Observation, Value, ValueWriter};
+use smallvec::SmallVec;
+use std::marker::PhantomData;
+
+/// Strategy for aggregating observations in a histogram.
+///
+/// Implementations determine how values are stored and converted to observations
+/// when the histogram is closed.
+pub trait AggregationStrategy {
+    /// Record a single observation.
+    fn record(&mut self, value: f64);
+
+    /// Drain all observations and return them as a vector.
+    ///
+    /// This resets the strategy's internal state.
+    fn drain(&mut self) -> Vec<Observation>;
+}
+
+/// Thread-safe strategy for aggregating observations in a histogram.
+///
+/// Like [`AggregationStrategy`] but allows recording values through a shared reference.
+pub trait AtomicAggregationStrategy {
```

Comment:
Maybe this should be "SharedAggregationStrategy" rather than "Atomic"? You can have a shared non-atomic aggregated strategy (using `RefCell` or the like) using this trait